File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.go
package main

import (
	"bufio"
	"os"
	"fmt"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)

	// Read 1st line
	n, m := getTwoIntsFromStdin(in)

	// Init graph
	graph := make(map[int][]int)
	visited := make(map[int]bool)

	for i := 1; i <= m; i++ {
		x, y := getTwoIntsFromStdin(in)
		if nodes, ok := graph[x]; ok {
			graph[x] = append(nodes, y)
		} else {
			visited[x] = false
			graph[x] = []int{y}
		}

		if nodes, ok := graph[y]; ok {
			graph[y] = append(nodes, x)
		} else {
			visited[y] = false
			graph[y] = []int{x}
		}
	}

	cnt := 0

	// Traverse
	for i := 1; i <= n; i++ {
		if !visited[i] {
			traverseFromNode(i, graph, visited)
			cnt += 1
		}
	}

	out.WriteString(fmt.Sprintf("%d\n", cnt))
	out.Flush()
}

func traverseFromNode(node int, graph map[int][]int, visited map[int]bool) {
	next := graph[node]
	visited[node] = true

	for _, n := range next {
		if !visited[n] {
			traverseFromNode(n, graph, visited)
		}
	}
}

func getTwoIntsFromStdin(in *bufio.Reader) (int, int) {
	var n, m int

	s, _ := in.ReadString('\n')
	fmt.Sscanf(s, "%d %d", &n, &m)

	return n, m
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
from typing import Set, List

# 1st line
n, m = map(
        lambda s: int(s),
        sys.stdin.readline().split()
    )

# cases
graph: List[Set[int]] = [None]
notVisited: Set[int] = set()
cnt = 0

for i in range(1, n+1):
    graph.append(set())
    notVisited.add(i)

for _ in range(m):
    x, y = map(
            lambda s: int(s),
            sys.stdin.readline().split()
        )
    graph[x].add(y)
    graph[y].add(x)

def moveThrough(i: int):
    next = graph[i]
    toVisit = next.intersection(notVisited)
    notVisited.difference_update(toVisit)
    for v in toVisit:
        moveThrough(v)


for i in range(1, n+1):
    if i not in notVisited:
        continue

    moveThrough(i)

    cnt += 1

sys.stdout.write(f"{cnt}")
